rails server: Allow to explicitly specify whether to output Rails's log to stdout

doits opened the pull request:
### Summary

Allow to explicitly specify whether to output Rails's log to stdout

Before Rails' logger output is mirrored to std out if:

	* environment is development and
	* the process is not daemonized

It was not possible to change that behaviour, e.g. to disable log output
in that case or enable it in other cases.

Now you can explicitly disable or enable output with the new command
line switch `-l`, regardless of any other circumstances.

```
// enable output even in production
rails server -e production -l

// disable output in development
rails server -e development -l false
```

Enabling output when daemonized still makes no sense (since tty is
detached), so the flag is ignored then and daemonizing always disables
output.

If the command line flag `-l` is not specified, the old behaviour still
applies, so this change is completely backward compatible.

doits commented:
edit: right after submitting the patch it came to my mind that outputting something while daemonized totally makes no sense, so I updated this PR accordingly

gmcgibbon commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -112,12 +112,18 @@ class ServerCommand < Base # :nodoc:
         desc: "Specifies the PID file."
       class_option "dev-caching", aliases: "-C", type: :boolean, default: nil,
         desc: "Specifies whether to perform caching in development."
+      class_option "log-stdout", aliases: "-l", type: :boolean, default: nil,
```
Should the alias be `-s` to signify output is going to `STDOUT` or maybe `-v` for verbose? `-l` is good too but I'd like to hear some other opinions on this.

kaspth replied:
I don't think this has enough use to need a short hand.

doits replied:
I just removed the shorthand.

gmcgibbon commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -112,12 +112,18 @@ class ServerCommand < Base # :nodoc:
         super
         @server = self.args.shift
-        @log_stdout = options[:daemon].blank? && (options[:environment] || Rails.env) == "development"
+        @log_stdout =
```
Maybe this should be moved to a memoized private function named something like `log_to_stdout?`? I think with the logic change this makes sense.

kaspth replied:
Seems good. I wouldn't even memoize it. It's not a candidate for being called thousands of times, so the 2-3 calls in a server boot are fine to redo the work.

doits replied:
Alright, removed the memoizing.

gmcgibbon reviewed (commented):
Still an issue on Rails 6 (a0061d2389a178b093f0d3f64f58236ffbe088e0). I can think of a few instances where I wanted to log output to `STDOUT` in a different environment but couldn't so 👍

kaspth commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -112,12 +112,18 @@ class ServerCommand < Base # :nodoc:
         desc: "Specifies the PID file."
       class_option "dev-caching", aliases: "-C", type: :boolean, default: nil,
         desc: "Specifies whether to perform caching in development."
+      class_option "log-stdout", aliases: "-l", type: :boolean, default: nil,
```
Thor automatically dasherizes options for the CLI. So let's call this `:log_stdout`, which makes it easier to access in `options` as well.

doits replied:
👍 After rebasing I saw this is used for other options as well, so amended it.

kaspth commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -112,12 +112,18 @@ class ServerCommand < Base # :nodoc:
       class_option "dev-caching", aliases: "-C", type: :boolean, default: nil,
         desc: "Specifies whether to perform caching in development."
+      class_option "log-stdout", aliases: "-l", type: :boolean, default: nil,
+        desc: "Specifies whether to output Rails' log to std out (true = always, false = never, default = in development). Running as a Daemon always disables output."
```
There's something about having a boolean mean 3 things that seems off to me. I'd think we'd only want `--log-to-stdout` (e.g. no `--no-log-to-stdout`) which just overrides whatever logging setup is there. Perhaps we should just honor it despite `:daemon` as well? Or throw an exception if both are enabled?

doits replied:
Changed it so that it raises in case of being daemonized and logging enabled.

It is still possible to do `--log-to-stdout false` and "--no-log-to-stoud" afaik though, which is OK imo.

doits commented:
rebased and changed after feedback. I can squash into one commit if everything is ok for you.

kaspth commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -132,13 +134,14 @@ class ServerCommand < Base # :nodoc:
       class_option :restart, type: :boolean, default: nil, hide: true
       class_option :early_hints, type: :boolean, default: nil, desc: "Enables HTTP/2 early hints."
+      class_option :log_to_stdout, type: :boolean, default: nil,
+        desc: "Specifies whether to output Rails' log to std out (true = always, false = never, default = in development)."
```
Think we can tune this wording a bit:

```ruby
class_option :log_to_stdout, type: :boolean, optional: true,
  desc: "Whether to log to stdout. Enabled by default in development when not daemonized."
```

kaspth commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -86,6 +86,8 @@ def create_tmp_directories
       end
 
       def log_to_stdout
+        raise "cannot log to stdout if daemonized" if options[:daemonize]
```
Let's just skip this, once the app daemonizes it's pretty easy to see that nothing will be written to stdout. The option is misspelled too.

kaspth commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -256,6 +259,12 @@ def early_hints
+        def log_to_stdout?
+          return options[:log_to_stdout] unless options[:log_to_stdout].nil?
+
+          options[:daemon].blank? && environment == "development"
```
Let's go with:

```ruby
def log_to_stdout?
  options.fetch(:log_to_stdout) do
    options[:daemon].blank? && environment == "development"
  end
end
```

That, coupled with the `optional: true, default: nil` in the `class_option` should make it such that `--log-to-stdout` or `--no-log-to-stdout` always overrides.

kaspth commented:
Yes, please squash your commits down to one 👍 — sorry for going so many rounds on this, it's been a while since I've last really touched the server command.

doits committed:
Allow to explicitly specify whether to output Rails' log to stdout

Before Rails' logger output is mirrored to std out if:

* environment is development and
* the process is not daemonized

It was not possible to change that behaviour, e.g. to disable log output
in that case or enable it in other cases.

Now you can explicitly disable or enable output with the new command
line switch `--log-to-stdout`, regardless of any other circumstances.

```
// enable output in production
rails server -e production --log-to-stdout

// disable output in development
rails server -e development --no-log-to-stdout
```

Enabling output when daemonized still makes no sense (since tty is
detached), but this is ignored for now.

If the command line flag is not specified, old behaviour still
applies, so this change is completely backward compatible.

doits commented:
@kaspth thanks for review, I amended everything. Hopefully tests will pass 🤞 

And np for going some rounds: We both want good code in the end, sometimes it takes some rounds. And it was my first time to touch the server command, so I didn't expect it to be bulletproof on first try anyway.

kaspth commented:
That's exactly the right attitude! Thanks, @doits 👏
